### ISSUE TITLE
Deadlock in VFS after "cancelling writeback"

### ISSUE THREAD

**Issue by esticle** [rclone/rclone#5355]
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/vfs-cache-stale-files-in-sftp-mode/24372

#### What is the problem you are having with rclone?

It appears that when the debug logs indicate 'cancelling writeback' the vfs cache stops reporting every minute, directory listings become stale, uploads still occur however gets 'stuck' at 100%

#### What is your rclone version (output from `rclone version`)

v1.55.1

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Container rclone/rclone:latest on Ubuntu 64

#### Which cloud storage system are you using? (e.g. Google Drive)

Google Drive

#### The command you were trying to run 

> rclone --config /config/rclone .conf serve sftp --authorized-keys=/config/authorized_keys --dir-cache-time=1000h --buffer-size=128M --addr=:2022 --log-level=DEBUG --log-file=/log/rclone.log --user-agent=mycustom1 --cache-dir=/data/cache --vfs-cache-mode=full --vfs-read-ahead=256M --vfs-cache-max-age=300h --vfs-cache-max-size=5G gdrive2-data:.users/brick 

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

n normal operations the DEBUG log gets updated every minute with a 'vfs cache: cleaned:' line - I can pin-point the time the cache "stops working" roughly here:

> $ grep uploading rclone.log  | tail -5
> 2021/05/21 00:16:50 INFO  : vfs cache: cleaned: objects 144 (was 144) in use 17, to upload 6, uploading 2, total size 4.627G (was 4.627G)
> 2021/05/21 00:17:50 INFO  : vfs cache: cleaned: objects 191 (was 191) in use 11, to upload 2, uploading 0, total size 4.760G (was 4.760G)
> 2021/05/21 00:18:50 INFO  : vfs cache: cleaned: objects 241 (was 241) in use 17, to upload 4, uploading 3, total size 4.893G (was 4.893G)
> 2021/05/21 00:19:50 INFO  : vfs cache: cleaned: objects 247 (was 292) in use 17, to upload 8, uploading 0, total size 4.998G (was 5.026G)
> 2021/05/21 00:20:50 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fd71d6f01e399672770dca83aeed935e.png: vfs cache: cancelling writeback (uploading true) 0xc00564cf50 item 19325
> 

It appears that 'vfs cache: cleaned:' stops reporting every minute once 'vfs cache: cancelling writeback' occurs.

Whilst in this state on the client side (using a sftp client) some directories list, others not:

> $ sftp -P2022 server
> Connected to server
> sftp> ls
> backedup-storage    backups             backup2             
> sftp> ls -al
> drwxr-xr-x    0 root     root            0 May 18 20:30 backedup-storage
> drwxr-xr-x    0 root     root            0 May 18 20:35 backups
> drwxr-xr-x    0 root     root            0 May 18 20:31 backups2
> sftp> cd backups
> sftp> ls -al
> **times out**

Once I restart rclone (the container in my case) - it appears that all the content not yet uploaded on the vfs cache gets uploaded, and everything returns to normal, until 'vfs cache: cancelling writeback' occurs in the logs again.

#### Whilst in the error state

- The sftp client can upload others files, however they remain stuck at '100%', the logs then say:

> 2021/05/21 07:51:48 DEBUG : backedup-storage/Videos/VHS/Las_Loop_Modeparade.mkv(0xc003dcb780): writeAt: size=24114, off=10707628032
> 2021/05/21 07:51:48 DEBUG : backedup-storage/Videos/VHS/Las_Loop_Modeparade.mkv(0xc003dcb780): >_writeAt: n=24114, err=<nil>
> 2021/05/21 07:51:48 DEBUG : backedup-storage/Videos/VHS/Las_Loop_Modeparade.mkv(0xc003dcb780): close:
> 2021/05/21 07:51:48 DEBUG : backedup-storage/Videos/VHS/Las_Loop_Modeparade.mkv: vfs cache: set ting modification time to 2021-05-21 07:51:48.308072334 +0000 UTC m=+57963.037840797
> 2021/05/21 07:51:48 INFO  : backedup-storage/Videos/VHS/Las_Loop_Modeparade.mkv: vfs cache: queuing for upload in 5s
> 2021/05/21 07:51:48 DEBUG : backedup-storage/Videos/VHS/Las_Loop_Modeparade.mkv(0xc003dcb780): >close: err=<nil>
> 

- These files (such as the above) are not remaining open (checked lsof), as expected
- These files has the correct filesize within the vfs cache-dir md5sum's of the file on the vfs cache-dir and source material also match
- As mentioned, a simple rclone restart will ensure this all gets read-again and uploads correctly and all is normal under we hit the error condition again

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
This is where the problem happened. The symptoms are caused by the directory lock being held and not released.

```
1: chan receive [640 minutes] [Created by vfscache.New @ cache.go:144]
    writeback  writeback.go:385      (*WriteBack)._cancelUpload(*WriteBack(#90), *writeBackItem(#1153))
    writeback  writeback.go:291      (*WriteBack).Remove(*WriteBack(#90), Handle(0xdb04))
    vfscache   item.go:874           (*Item)._remove(*Item(#795), string(#9, len=34), 0x5a96)
    vfscache   item.go:923           (*Item).RemoveNotInUse(*Item(#795), 0x0, #859, 0x1, 0x5a96)
    vfscache   cache.go:502          (*Cache).removeNotInUse(*Cache(#92), *Item(#795), Duration(0x0))
    vfscache   cache.go:652          (*Cache).purgeOverQuota(*Cache(#92), 5368709120)
    vfscache   cache.go:679          (*Cache).clean(*Cache(#92), bool(#841))
    vfscache   cache.go:742          (*Cache).cleaner(#92, #26, #311)
```

So you are right it happened when cancelling the upload.

However there are two things I don't understand yet

1. This appears to be running the cache purge so cleaning your cache back down to size. Looking at the log below the file appears to have been uploaded correctly but somehow the vfs cache still appears to want to cancel the upload

2. This caused a deadlock somehow - the cancel for the upload never returns (probably because the file wasn't uploading) so the above trace waits forever for it.

Here are some more logs for the file in question (from the forum) - the deadlock happened at the "cancelling upload" log line.

```
2021/05/21 22:24:32 DEBUG : /backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: OpenFile: flags=O_WRONLY|O_CREATE|O_TRUNC, perm=-rwxrwxrwx
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: Open: flags=O_WRONLY|O_CREATE|O_TRUNC
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: newRWFileHandle: 
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png(0xc00ae55880): openPending: 
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: vfs cache: truncate to size=0
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png(0xc00ae55880): >openPending: err=<nil>
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: >newRWFileHandle: err=<nil>
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: >Open: fd=backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png (rw), err=<nil>
2021/05/21 22:24:32 DEBUG : /backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: >OpenFile: fd=backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png (rw), err=<nil>
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png(0xc00ae55880): _writeAt: size=23190, off=0
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png(0xc00ae55880): >_writeAt: n=23190, err=<nil>
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png(0xc00ae55880): _writeAt: size=0, off=23190
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png(0xc00ae55880): >_writeAt: n=0, err=<nil>
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png(0xc00ae55880): close: 
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: vfs cache: setting modification time to 2021-05-21 22:24:32.428790008 +0000 UTC m=+37680.681425438
2021/05/21 22:24:32 INFO  : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: vfs cache: queuing for upload in 5s
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png(0xc00ae55880): >close: err=<nil>
2021/05/21 22:24:32 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: vfs cache: setting modification time to 2000-12-08 00:14:37 +0000 UTC
2021/05/21 22:24:33 DEBUG : serve sftp 102.65.40.41:41870->172.19.0.14:2022: exec command: binary = "md5sum", args = "backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png"
2021/05/21 22:24:33 DEBUG : serve sftp 102.65.40.41:41870->172.19.0.14:2022: command "md5sum backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png" failed with error: unexpected non file
2021/05/21 22:24:37 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: vfs cache: starting upload
2021/05/21 22:24:38 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: MD5 = dd26141b5189d019869390febcba7503 OK
2021/05/21 22:24:38 INFO  : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: Copied (new)
2021/05/21 22:24:38 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: vfs cache: fingerprint now "23190,2000-12-08 00:14:37 +0000 UTC"
2021/05/21 22:24:38 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: vfs cache: writeback object to VFS layer
2021/05/21 22:24:38 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: Not setting pending mod time 2000-12-08 00:14:37 +0000 UTC as it is already set
2021/05/21 22:25:20 DEBUG : vfs cache RemoveNotInUse (maxAge=1080000000000000, emptyOnly=false): item backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png not removed, freed 0 bytes
2021/05/21 22:25:20 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: vfs cache: cancelling writeback (uploading true) 0xc031e88cb0 item 56068
2021/05/21 22:25:20 DEBUG : backups/stasiewa/marais/.thumbnails/normal/fe1f1749570c2d75a34bd8af9646c0fb.png: vfs cache: cancelling upload
```

**Comment by ncw** [comment]
I think the sequence of events went like this. **1** and **2** are the different threads

1. **1**: rclone scans the cache for old files holding the `Cache` mutex
2. **2**: an upload starts and finishes except for the adding to the VFS in memory `Dir`
3. **2**: this is blocked on the `Cache` mutex so needs to wait for the cache scan to complete. It is holding the `Dir` mutex here.
4. **1**: the cache scan finds a file it wants to remove (the same file as above)
5. **1**: the upload is still in progress since it hasn't finished the final VFS write, though the the data is now clean
6. **1**: Trying to remove it blocks forever since the VFS writeback is blocked on the `Cache` mutex.

Here is the edited and annotated deadlock trace

[issue-5355-vfs-deadlock-parsed-edit.txt](https://github.com/rclone/rclone/files/6532458/issue-5355-vfs-deadlock-parsed-edit.txt)

So, how to fix this...

Doing 2) in a separate go routine would fix the deadlock. This would make the update asynchronous to the file writeback.  I tried this and unfortunately it breaks the `--vfs-writeback 0` expectations resulting in lots of test failures :-(

Expiring the file without the Cache lock being held in 1) might be sensible also. Or turning the cache lock into a RWLock and holding RLock only while scanning the cache.

**Comment by ncw** [comment]
I think the internals of the VFS have changed so much since this issue was made that it is no longer relevant.
